In [163]:
import pandas as pd
import numpy as np

In [164]:
df = pd.read_csv("2020-XTern-DS.csv")
df.head(10)

Restaurant   Latitude  Longitude  \
0    ID_6321  39.262605 -85.837372   
1    ID_2882  39.775933 -85.740581   
2    ID_1595  39.253436 -85.123779   
3    ID_5929  39.029841 -85.332050   
4    ID_6123  39.882284 -85.517407   
5    ID_5221  39.370441 -85.739516   
6    ID_3777  39.821806 -85.005577   
7     ID_745  39.280324 -85.144363   
8    ID_2970  39.268816 -85.602168   
9    ID_3474  39.874521 -85.439963   

                                            Cuisines Average_Cost  \
0             Fast Food, Rolls, Burger, Salad, Wraps       $20.00   
1                                Ice Cream, Desserts       $10.00   
2                    Italian, Street Food, Fast Food       $15.00   
3                     Mughlai, North Indian, Chinese       $25.00   
4                                    Cafe, Beverages       $20.00   
5                South Indian, North Indian, Chinese       $15.00   
6                               Beverages, Fast Food       $15.00   
7                               Chinese, Thai, Asian       $65.00   
8                                Mithai, Street Food       $10.00   
9  Fast Food, North Indian, Rolls, Chinese, Momos...       $20.00   

  Minimum_Order Rating Votes Reviews   Cook_Time  
0        $50.00    3.5    12       4  30 minutes  
1        $50.00    3.5    11       4  30 minutes  
2        $50.00    3.6    99      30  65 minutes  
3        $99.00    3.7   176      95  30 minutes  
4        $99.00    3.2   521     235  65 minutes  
5        $50.00    3.8    46      18  30 minutes  
6        $50.00    3.7   108      31  30 minutes  
7        $50.00    4.0  1731    1235  45 minutes  
8        $50.00    3.9   110      26  30 minutes  
9        $50.00    3.9   562     294  65 minutes

In [165]:
list(df.columns) #list(dating_data)

['Restaurant',
 'Latitude',
 'Longitude',
 'Cuisines',
 'Average_Cost',
 'Minimum_Order',
 'Rating',
 'Votes',
 'Reviews',
 'Cook_Time']

In [166]:
df.isnull().any()

Restaurant       False
Latitude         False
Longitude        False
Cuisines         False
Average_Cost     False
Minimum_Order    False
Rating           False
Votes            False
Reviews          False
Cook_Time        False
dtype: bool

In [167]:
# remove $ from 3 columns
df["Average_Cost"] = df["Average_Cost"].str.replace('$', '')
df["Average_Cost"] = df["Average_Cost"].str.replace(',', '.')
df["Minimum_Order"] = df["Minimum_Order"].str.replace('$', '')
df["Cook_Time"] = df["Cook_Time"].str.replace(' minutes', '')
df.head(10)

Restaurant   Latitude  Longitude  \
0    ID_6321  39.262605 -85.837372   
1    ID_2882  39.775933 -85.740581   
2    ID_1595  39.253436 -85.123779   
3    ID_5929  39.029841 -85.332050   
4    ID_6123  39.882284 -85.517407   
5    ID_5221  39.370441 -85.739516   
6    ID_3777  39.821806 -85.005577   
7     ID_745  39.280324 -85.144363   
8    ID_2970  39.268816 -85.602168   
9    ID_3474  39.874521 -85.439963   

                                            Cuisines Average_Cost  \
0             Fast Food, Rolls, Burger, Salad, Wraps        20.00   
1                                Ice Cream, Desserts        10.00   
2                    Italian, Street Food, Fast Food        15.00   
3                     Mughlai, North Indian, Chinese        25.00   
4                                    Cafe, Beverages        20.00   
5                South Indian, North Indian, Chinese        15.00   
6                               Beverages, Fast Food        15.00   
7                               Chinese, Thai, Asian        65.00   
8                                Mithai, Street Food        10.00   
9  Fast Food, North Indian, Rolls, Chinese, Momos...        20.00   

  Minimum_Order Rating Votes Reviews Cook_Time  
0         50.00    3.5    12       4        30  
1         50.00    3.5    11       4        30  
2         50.00    3.6    99      30        65  
3         99.00    3.7   176      95        30  
4         99.00    3.2   521     235        65  
5         50.00    3.8    46      18        30  
6         50.00    3.7   108      31        30  
7         50.00    4.0  1731    1235        45  
8         50.00    3.9   110      26        30  
9         50.00    3.9   562     294        65

In [168]:
df.loc[df['Average_Cost'] =='-','Average_Cost']=np.nan
df.loc[df['Minimum_Order'] =='-','Minimum_Order']=np.nan
df.loc[df['Rating'] =='-','Rating']=np.nan
df.loc[df['Votes'] =='-','Votes']=np.nan
df.loc[df['Reviews'] =='-','Reviews']=np.nan

df = df.apply(pd.to_numeric, errors='ignore')
df.head()

Restaurant   Latitude  Longitude                                Cuisines  \
0    ID_6321  39.262605 -85.837372  Fast Food, Rolls, Burger, Salad, Wraps   
1    ID_2882  39.775933 -85.740581                     Ice Cream, Desserts   
2    ID_1595  39.253436 -85.123779         Italian, Street Food, Fast Food   
3    ID_5929  39.029841 -85.332050          Mughlai, North Indian, Chinese   
4    ID_6123  39.882284 -85.517407                         Cafe, Beverages   

   Average_Cost  Minimum_Order Rating  Votes  Reviews  Cook_Time  
0          20.0           50.0    3.5   12.0      4.0         30  
1          10.0           50.0    3.5   11.0      4.0         30  
2          15.0           50.0    3.6   99.0     30.0         65  
3          25.0           99.0    3.7  176.0     95.0         30  
4          20.0           99.0    3.2  521.0    235.0         65

In [169]:
df['Rating'].value_counts() 

3.7             169
3.5             149
3.8             141
NEW             135
3.9             131
3.3             131
3.4             130
3.6             124
4.0             113
4.1              89
3.2              87
3.1              80
4.2              53
3.0              51
4.3              45
2.9              43
2.8              42
4.4              27
2.7              13
4.5              12
4.7              10
2.6               9
4.6               9
2.5               5
Opening Soon      2
4.8               2
2.4               1
Name: Rating, dtype: int64

For 'NEW', 'Opening Soon', data cleaning

In [170]:
df.loc[df['Rating'] =='NEW','Rating']=np.nan
df.loc[df['Rating'] =='Opening Soon','Rating']=np.nan
df['Rating'] = df['Rating'].apply(pd.to_numeric)

Average Cook_Time is 36.92 mins and average Rating is 3.18
The mean of Latitude and Longitude is 39.50 and -85.50, which means the optimized FoodieX pick up zones would be around(39.50, -85.50)

In [171]:
df.describe()

Latitude    Longitude  Average_Cost  Minimum_Order       Rating  \
count  2019.000000  2019.000000   2019.000000    2019.000000  1666.000000   
mean     39.499131   -85.503494     20.015948      53.342249     3.609304   
std       0.285759     0.290395     12.684006      20.314896     0.422452   
min      39.000105   -85.998864      1.000000       0.000000     2.400000   
25%      39.263364   -85.748611     10.000000      50.000000     3.300000   
50%      39.494870   -85.509055     20.000000      50.000000     3.600000   
75%      39.741469   -85.245741     20.000000      50.000000     3.900000   
max      39.999523   -85.001541    150.000000     450.000000     4.800000   

             Votes      Reviews    Cook_Time  
count  1639.000000  1604.000000  2019.000000  
mean    254.799268   127.373441    36.921743  
std     561.674069   330.305363    11.804352  
min       4.000000     1.000000    10.000000  
25%      18.000000     7.000000    30.000000  
50%      65.000000    27.000000    30.000000  
75%     246.000000   102.250000    45.000000  
max    9054.000000  6504.000000   120.000000

Votes and reviews miss lots of information - 0.0

In [172]:
df['Votes'].value_counts() 

4.0       45
12.0      39
7.0       38
9.0       37
6.0       30
          ..
161.0      1
1552.0     1
648.0      1
155.0      1
910.0      1
Name: Votes, Length: 523, dtype: int64

In [173]:
df['Reviews'].value_counts() 

2.0       78
1.0       71
3.0       70
5.0       62
4.0       59
          ..
565.0      1
1150.0     1
518.0      1
938.0      1
2576.0     1
Name: Reviews, Length: 359, dtype: int64

In [175]:
df['Rating'].value_counts() 

3.7    169
3.5    149
3.8    141
3.3    131
3.9    131
3.4    130
3.6    124
4.0    113
4.1     89
3.2     87
3.1     80
4.2     53
3.0     51
4.3     45
2.9     43
2.8     42
4.4     27
2.7     13
4.5     12
4.7     10
4.6      9
2.6      9
2.5      5
4.8      2
2.4      1
Name: Rating, dtype: int64

In [176]:
# drop data point while have at least 2 NaN
df = df.dropna(thresh=2)

In [177]:
df.describe()

Latitude    Longitude  Average_Cost  Minimum_Order       Rating  \
count  2019.000000  2019.000000   2019.000000    2019.000000  1666.000000   
mean     39.499131   -85.503494     20.015948      53.342249     3.609304   
std       0.285759     0.290395     12.684006      20.314896     0.422452   
min      39.000105   -85.998864      1.000000       0.000000     2.400000   
25%      39.263364   -85.748611     10.000000      50.000000     3.300000   
50%      39.494870   -85.509055     20.000000      50.000000     3.600000   
75%      39.741469   -85.245741     20.000000      50.000000     3.900000   
max      39.999523   -85.001541    150.000000     450.000000     4.800000   

             Votes      Reviews    Cook_Time  
count  1639.000000  1604.000000  2019.000000  
mean    254.799268   127.373441    36.921743  
std     561.674069   330.305363    11.804352  
min       4.000000     1.000000    10.000000  
25%      18.000000     7.000000    30.000000  
50%      65.000000    27.000000    30.000000  
75%     246.000000   102.250000    45.000000  
max    9054.000000  6504.000000   120.000000

In [178]:
trend = df[[ 'Rating' , 'Votes', 'Restaurant' ]]
trend.head()

Rating  Votes Restaurant
0     3.5   12.0    ID_6321
1     3.5   11.0    ID_2882
2     3.6   99.0    ID_1595
3     3.7  176.0    ID_5929
4     3.2  521.0    ID_6123

In [179]:
trend = trend.groupby('Restaurant')
trend.mean()

Rating   Votes
Restaurant                
ID_1000        3.7    96.0
ID_1004        3.0     4.0
ID_1005        4.3   253.0
ID_1007        NaN     NaN
ID_1013        3.3     5.0
...            ...     ...
ID_986         NaN     NaN
ID_988         3.9  2768.0
ID_989         4.1  1177.0
ID_99          3.8   101.0
ID_996         NaN     NaN

[1807 rows x 2 columns]

In [180]:
trend.describe()

Rating                                   Votes                      \
            count mean std  min  25%  50%  75%  max count    mean std     min   
Restaurant                                                                      
ID_1000       1.0  3.7 NaN  3.7  3.7  3.7  3.7  3.7   1.0    96.0 NaN    96.0   
ID_1004       1.0  3.0 NaN  3.0  3.0  3.0  3.0  3.0   1.0     4.0 NaN     4.0   
ID_1005       1.0  4.3 NaN  4.3  4.3  4.3  4.3  4.3   1.0   253.0 NaN   253.0   
ID_1007       0.0  NaN NaN  NaN  NaN  NaN  NaN  NaN   0.0     NaN NaN     NaN   
ID_1013       1.0  3.3 NaN  3.3  3.3  3.3  3.3  3.3   1.0     5.0 NaN     5.0   
...           ...  ...  ..  ...  ...  ...  ...  ...   ...     ...  ..     ...   
ID_986        0.0  NaN NaN  NaN  NaN  NaN  NaN  NaN   0.0     NaN NaN     NaN   
ID_988        1.0  3.9 NaN  3.9  3.9  3.9  3.9  3.9   1.0  2768.0 NaN  2768.0   
ID_989        1.0  4.1 NaN  4.1  4.1  4.1  4.1  4.1   1.0  1177.0 NaN  1177.0   
ID_99         1.0  3.8 NaN  3.8  3.8  3.8  3.8  3.8   1.0   101.0 NaN   101.0   
ID_996        0.0  NaN NaN  NaN  NaN  NaN  NaN  NaN   0.0     NaN NaN     NaN   

                                            
               25%     50%     75%     max  
Restaurant                                  
ID_1000       96.0    96.0    96.0    96.0  
ID_1004        4.0     4.0     4.0     4.0  
ID_1005      253.0   253.0   253.0   253.0  
ID_1007        NaN     NaN     NaN     NaN  
ID_1013        5.0     5.0     5.0     5.0  
...            ...     ...     ...     ...  
ID_986         NaN     NaN     NaN     NaN  
ID_988      2768.0  2768.0  2768.0  2768.0  
ID_989      1177.0  1177.0  1177.0  1177.0  
ID_99        101.0   101.0   101.0   101.0  
ID_996         NaN     NaN     NaN     NaN  

[1807 rows x 16 columns]

In [181]:
# Take 75% as baseline
querySer1=df.loc[:,'Reviews']>102
data_review1=df.loc[querySer1,['Restaurant','Reviews','Average_Cost','Minimum_Order','Votes','Rating','Cook_Time']]
data_review1.sort_values(by='Votes',ascending=False)

Restaurant  Reviews  Average_Cost  Minimum_Order   Votes  Rating  \
1180    ID_1064   6504.0          80.0           50.0  9054.0     4.7   
1300    ID_1666   3989.0          50.0           99.0  4903.0     4.6   
123     ID_2885   2547.0          45.0           50.0  4691.0     4.3   
1780    ID_2601   2311.0          15.0           50.0  4606.0     4.1   
1796    ID_6511   2576.0          10.0           50.0  4438.0     3.9   
...         ...      ...           ...            ...     ...     ...   
1807    ID_7635    104.0          20.0           50.0   148.0     4.2   
1781    ID_6514    112.0          20.0           50.0   143.0     4.0   
130     ID_7674    110.0          25.0           99.0   143.0     4.5   
946     ID_4666    126.0          10.0           50.0   133.0     4.3   
137     ID_5401    109.0          15.0           50.0   124.0     4.2   

      Cook_Time  
1180         45  
1300         65  
123          45  
1780         45  
1796         45  
...         ...  
1807         30  
1781         30  
130          65  
946          30  
137          30  

[401 rows x 7 columns]

We can see the most popular restaurant is 1064, which achieve 4.7 rate within 6504 reviews, 9054 votes

In [182]:
querySer2=(data_review1.loc[:,'Rating']>=3.9)&(data_review1.loc[:,'Votes']>246)
data_review2=data_review1.loc[querySer2,:]
data_review2.sort_values(by='Votes',ascending=False)

Restaurant  Reviews  Average_Cost  Minimum_Order   Votes  Rating  \
1180    ID_1064   6504.0          80.0           50.0  9054.0     4.7   
1300    ID_1666   3989.0          50.0           99.0  4903.0     4.6   
123     ID_2885   2547.0          45.0           50.0  4691.0     4.3   
1780    ID_2601   2311.0          15.0           50.0  4606.0     4.1   
1796    ID_6511   2576.0          10.0           50.0  4438.0     3.9   
...         ...      ...           ...            ...     ...     ...   
148     ID_1005    118.0          15.0           50.0   253.0     4.3   
1494    ID_1056    105.0          15.0           50.0   251.0     3.9   
514     ID_2286    121.0          20.0           50.0   250.0     4.3   
579     ID_8062    139.0          35.0           50.0   249.0     4.3   
1987    ID_4325    178.0          20.0           50.0   247.0     4.5   

      Cook_Time  
1180         45  
1300         65  
123          45  
1780         45  
1796         45  
...         ...  
148          30  
1494         65  
514          30  
579          30  
1987         30  

[277 rows x 7 columns]